core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3380,6 +3384,66 @@ protected void validateFrom(
     getNamespaceOrThrow(node, scope).validate(targetRowType);
   }
 
+  protected void validateTableFunction(SqlCall node, SqlValidatorScope scope,
+      RelDataType targetRowType) {
+    // Dig out real call; TABLE() wrapper is just syntactic.
+    SqlCall call = node.operand(0);
+    if (call.getOperator() instanceof SqlTableFunction) {
+      SqlTableFunction tableFunction = (SqlTableFunction) call.getOperator();
```

Comment:
What kind of function it is if the SqlKind is `COLLECTION_TABLE` but not a `SqlTableFunction` ?

Reply:
Currently, there is no valid case like this.
It requires argument of COLLECTION_TABLE must be a table function which is already checked by ProcedureNamespace#validateImpl.
SqlValidatorTest#testCollectionTable cover these cases.

Reply:
So the `instance of ` is not necessary here ?

Reply:
I prefer to keep `instance of` here.
because `ProcedureNamespace#validateImpl` do a validation after this step.
If I remove `instance of` here, I need to add the `else` branch here to handle illegal cases.
WDYT?